File: libs/toolkit/src/main/java/com/example/toolkit/Optional.java
package com.example.toolkit;

import android.support.annotation.Nullable;

public class Optional<T> {
  private final T value;

  private Optional(T value) {
    this.value = value;
  }

  public static <T> Optional<T> from(T value) {
    if (value == null) {
      throw new NullPointerException();
    }

    return new Optional<>(value);
  }

  public static <T> Optional<T> fromNullable(@Nullable T value) {
    return new Optional<>(value);
  }

  public static <T> Optional<T> empty() {
    return new Optional<>(null);
  }

  public boolean isPresent() {
    return value != null;
  }

  public boolean isNotPresent() {
    return value == null;
  }

  public T get() {
    return value;
  }

  public T getOrDefault(T defaultValue) {
    if (isPresent()) {
      return value;
    }

    return defaultValue;
  }
}

File: libs/toolkit/src/main/kotlin/br/com/vitorsalgado/example/toolkit/Optional.kt
package br.com.vitorsalgado.example.toolkit

class Optional<T> private constructor(private val value: T?) {
  val isPresent: Boolean
    get() = value != null

  val isNotPresent: Boolean
    get() = value == null

  fun get(): T? {
    return value
  }

  fun getOrDefault(defaultValue: T): T? {
    return if (isPresent) {
      value
    } else defaultValue
  }

  companion object {
    fun <T> from(value: T?): Optional<T> {
      if (value == null) {
        throw NullPointerException()
      }

      return Optional(value)
    }

    fun <T> fromNullable(value: T?): Optional<T> {
      return Optional(value)
    }

    fun <T> empty(): Optional<T> {
      return Optional(null)
    }
  }
}
